### PR TITLE
Update C# codegen

### PR DESCRIPTION
**Pull request by chrisdunelm**
Uses immutable settings classes.
Fixes snippet bugs.
Add missing XML docs - updated #607.

### CODE REVIEW

**Comment by chrisdunelm** [comment]
This doesn't fix the ambiguous XML references, only the missing XML docs.

**Comment by chrisdunelm** [comment]
The generated client and snippet code all compiles without errors.

File: src/main/resources/com/google/api/codegen/csharp/gapic_client.snip [hunk 1]
```diff
@@ -457,6 +460,9 @@
         public partial class {@method.listMethod.requestTypeName} : IPageRequest { }
         public partial class {@method.listMethod.responseTypeName} : IPageResponse<{@method.listMethod.resourceTypeName}>
         {
+            /// <summary>
+            /// Returns an enumerator that iterates through the resources in this response.
+            /// </summary>
             public IEnumerator<{@method.listMethod.resourceTypeName}> GetEnumerator() => {@method.listMethod.resourcesFieldGetFunction}.GetEnumerator();
             IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
```
Comment (jskeet) [inline src/main/resources/com/google/api/codegen/csharp/gapic_client.snip]: We should have something on here as well - just `/// <inheritdoc/>` would be fine, I think.

File: src/test/java/com/google/api/codegen/testdata/csharp_gapic_client_library.baseline [hunk 2]
```diff
@@ -189,16 +185,13 @@ namespace Google.Example.Library.V1
...
-        {
-            Timing = CallTiming.FromRetry(new RetrySettings
-            {
-                RetryBackoff = GetDefaultRetryBackoff(),
-                TimeoutBackoff = GetDefaultTimeoutBackoff(),
-                RetryFilter = IdempotentRetryFilter,
-                TotalExpiration = Expiration.FromTimeout(TimeSpan.FromMilliseconds(30000)),
-            }),
-        };
+        public CallSettings GetShelfSettings { get; set; } = CallSettings.FromCallTiming(
```
Comment (jskeet) [inline src/test/java/com/google/api/codegen/testdata/csharp_gapic_client_library.baseline]: I've been wondering about a CallSettings.FromRetry, given how often we're likely to get this chaining.
We can add it later in a non-breaking way though.

File: src/main/resources/com/google/api/codegen/csharp/gapic_client.snip (hunk 1 above)
Comment (chrisdunelm) [inline src/main/resources/com/google/api/codegen/csharp/gapic_client.snip]: Done

File: src/test/java/com/google/api/codegen/testdata/csharp_gapic_client_library.baseline (hunk 2 above)
Comment (chrisdunelm) [inline src/test/java/com/google/api/codegen/testdata/csharp_gapic_client_library.baseline]: ack

**Pull request merged**